cranelift/codegen/src/isle_prelude.rs

```diff
@@ -343,6 +349,17 @@ macro_rules! isle_common_prelude_methods {
             imm.bits() as u64
         }
 
+        #[inline]
+        fn imm64_is_power_of_two(&mut self, x: Imm64) -> Option<u64> {
+            let x = i64::from(x);
+            let x = u64::try_from(x).ok()?;
+            if x.is_power_of_two() {
+                Some(x)
```

Comment:
I'd be inclined to return `x.trailing_zeros()` here instead of just `x`, and remove the `u64_log2` extractor. When you're asking whether something's a power of two, it's natural to wonder which power of two it is. On the flip side, using `trailing_zeros` to implement `ilog2` is safe _iff_ `x.is_power_of_two()` is true. So I think it makes the most sense to couple these two queries together pretty tightly. If somebody really just wants the original value after checking that it's a power of two, they can use at-patterns to bind the input instead of the result.

I also wonder, should this extractor take `u64` as input instead of `Imm64`? We have extractors for converting in either direction so I guess it doesn't matter too much which way we pick here.

Reply:
Yeah, I like that!

> I also wonder, should this extractor take u64 as input instead of Imm64? We have extractors for converting in either direction so I guess it doesn't matter too much which way we pick here.

Its just more wordy since we don't have automatic conversions between the two.

Reply:
(And I don't think we necessarily should since `Imm64` is "signed" but also it kind of depends on how the value ends up being used.)